### PR TITLE
Cross compilation

### PR DESCRIPTION
**Pull request by jpakkane**
This patchset makes the cross compilation work properly. It adds:

- the concept of build, host and target
- full canadian cross support
- target-only builds that use the host compilers (for compiling a cross-compiler)

It has only been tested with the unit tests, not with actual projects. I don't have those handy. It would be nice to get feedback from actual cross compilation people before merging this.

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
This branch has conflicts that must be resolved

File: interpreter.py [hunk 1]
```diff
@@ -1664,7 +1700,7 @@ def build_target(self, node, args, kwargs, targetholder):
     def check_sources_exist(self, subdir, sources):
         for s in sources:
             if not isinstance(s, str):
-                continue # This means a generated source and they always exist.
+                continue  # This means a generated source and they always exist.
```
Comment (ignatenkobrain) [inline interpreter.py]: unnecessary

File: interpreter.py [hunk 2]
```diff
@@ -1537,7 +1573,7 @@ def func_configure_file(self, node, args, kwargs):
         elif 'command' in kwargs:
             res = self.func_run_command(node, kwargs['command'], {})
             if res.returncode != 0:
-                raise InterpreterException('Running configure command failed.\n%s\n%s' %
+                raise InterpreterException('Running configure command failed.\n%s\n%s' %
```
Comment (ignatenkobrain) [inline interpreter.py]: not needed

File: interpreter.py [hunk 3]
```diff
@@ -1746,7 +1782,7 @@ def format_string(self, templ, args):
             args = args.arguments
         for (i, arg) in enumerate(args):
             arg = self.to_native(self.evaluate_statement(arg))
-            if isinstance(arg, bool): # Python boolean is upper case.
+            if isinstance(arg, bool):  # Python boolean is upper case.
```
Comment (ignatenkobrain) [inline interpreter.py]: not needed

File: interpreter.py [hunk 4]
```diff
@@ -170,7 +170,7 @@ def __init__(self, subdir, sourcename, targetname, configuration_data):
 class ConfigurationDataHolder(InterpreterObject):
     def __init__(self):
         super().__init__()
-        self.used = False # These objects become immutable after use in configure_file.
+        self.used = False  # These objects become immutable after use in configure_file.
```
Comment (ignatenkobrain) [inline interpreter.py]: not needed

File: cross/ubuntu-mingw.txt [hunk 5]
```diff
@@ -1,8 +1,22 @@
...
+[properties]
 root = '/usr/i686-w64-mingw32'
+
+[host_machine]
+name = 'windows'
+cpu = 'x86'
+endian = 'little'
+
+[target_machine]
+name='darwin'
```
Comment (ignatenkobrain) [inline cross/ubuntu-mingw.txt]: space before and after `=` ;) just to keep one style. next string also

File: interpreter.py [hunk 6]
```diff
@@ -920,7 +954,7 @@ def evaluate_codeblock(self, node):
                     e.colno = cur.colno
                     e.file = os.path.join(self.subdir, 'meson.build')
                 raise e
-            i += 1 # In THE FUTURE jump over blocks and stuff.
+            i += 1  # In THE FUTURE jump over blocks and stuff.
```
Comment (ignatenkobrain) [inline interpreter.py]: not needed

File: interpreter.py [hunk 7]
```diff
@@ -1021,7 +1055,7 @@ def evaluate_statement(self, cur):
     def validate_arguments(self, args, argcount, arg_types):
         if argcount is not None:
             if argcount != len(args):
-                raise InvalidArguments('Expected %d arguments, got %d.' %
+                raise InvalidArguments('Expected %d arguments, got %d.' %
```
Comment (ignatenkobrain) [inline interpreter.py]: not needed

**Comment by ignatenkobrain** [comment]
Looks good

**Comment by jpakkane** [comment]
retest this please

**Comment by jpakkane** [comment]
Where the hell do those indentation changes come from? I **definitely** am not doing them.

**Comment by ignatenkobrain** [comment]
I've pushed some cleanup commits. if cross-compilation works for you - feel free to merge.

**Pull request merged**